LowPtElectrons: final energy regression and ID

bainbrid opened the pull request:
#### PR description:

This PR contains:
- code that performs the energy regression as part of the default low-pT electrons RECO sequence 
- run2_miniAOD_UL modifier: schedules the energy regression prior to the PAT sequence (for existing samples)
- a switch to a new BDT model for the electron ID, "2020Nov28"
- selectedPatLowPtElectrons: BDT score threshold tuned to -0.25 to give 10% mistag rate

Studies on the energy regression and ID can be found [here](https://indico.cern.ch/event/983008/contributions/4140140/) and [here](https://indico.cern.ch/event/983567/contributions/4142431/). 

This PR depends on https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/16

For the UL re-miniAOD campaign, the 2020Sept15 model is used as part of the PAT sequence to select interesting electron candidates (PR https://github.com/cms-sw/cmssw/pull/32372, merged into CMSSW_10_6_20). This PR introduces the new 2020Nov28 model. The following slide provides details on how correlated the two BDT model scores are, and the effect on the 2020Nov28 performance after the application of a threshold of 1.5 on the score from the 2020Sept15 model. These studies are pertinent to e.g. a "UL nanoAOD v9" production based on existing UL re-miniAOD samples. [201223_ID_correlation.pdf](https://github.com/cms-sw/cmssw/files/5774997/201223_ID_correlation.pdf)

#### PR validation:

The new ID has been tested locally with wfs 1325.518 (TTbar reminiAOD) and 136.88811 (JetHT, 2018D) and a comparison made with scores from the previous "2020Sept15" model. The BDT scores are in line with expectations.

The energy regression has been tested with the following wfs and candidate GTs from EGamma:
- 1325.518 (TTbar reminiAOD), tested with 112X_upgrade2018_realistic_Candidate_2020_12_03_11_20_37
- 136.88811 (JetHT, 2018D), tested with 112X_dataRun2_relval_Candidate_2020_12_04_15_09_56

The regressed energies are in line with expectations.

bainbrid committed:
lowPtGsfElectron energy regression code

bainbrid committed:
clone from regressionModifier106XUL

bainbrid committed:
switch to fixedGridRhoFastjetAll, remove unchanged parameters

bainbrid committed:
add fillDescriptions

bainbrid committed:
switch to 2020Nov28 model trained with 10_6_X UL samples

bainbrid committed:
tune threshold to give 10% mistag rate

bainbrid commented:
@slava77 @crovelli @cavalari @jainshilpi @lsoffi @guitargeek

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32391/20257

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @bainbrid  for master.

It involves the following packages:

PhysicsTools/PatAlgos
RecoEgamma/EgammaElectronProducers


@perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @gouskos, @jainshilpi, @hatakeyamak, @emilbols, @varuns23, @seemasharmafnal, @mmarionncern, @ahinzmann, @smoortga, @lgray, @jdolen, @ferencek, @Sam-Harper, @rovere, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @andrzejnovak, @clelange, @JyothsnaKomaragiri, @sobhatta, @lecriste, @afiqaize, @gpetruc, @mariadalfonso, @ram1123 this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
test parameters:
- pull_request = cms-data/RecoEgamma-ElectronIdentification#16
- workflows = 136.898,1304.182

slava77 commented:
@cmsbuild please test

slava77 commented:
for the descoping scenarios, do I understand correctly that this PR (regression and ID code) can be used on top of miniAOD?

slava77 commented:
> for the descoping scenarios, do I understand correctly that this PR (regression and ID code) can be used on top of miniAOD?

if yes, is there a configuration test for this, I do not clearly see one in the PR itself.

bainbrid commented:
We can in principle apply energy regression and ID using MINIAOD inputs. However, this _should_ be based on a corresponding training using MINIAOD inputs. 

This PR assumes regression and ID are applied as part of RECO only. As such, the trainings are based on AOD. This is highly desirable in our opinion, as:
- it ensures the end user obtains a corrected P4 (otherwise he/she must use only the GsfTrack momentum mode)
- it permits to select the most signal-like electrons using the ID in the PAT selector step to maintain an acceptable event size.

cmsbuild commented:
-1

CMSSW: CMSSW_11_3_X_2020-12-04-1100
SCRAM_ARCH: slc7_amd64_gcc900
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-603162/11344/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>1001.0 step2
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVDSIPIXELCALRUN1+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/step2_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVDSIPIXELCALRUN1+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.log

</pre>1000.0 step2
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step2_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>136.88811 step2
<pre>runTheMatrix-results/136.88811_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL/step2_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL.log

</pre>140.56 step2
<pre>runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18.log

</pre>4.53 step3
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT/step3_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT.log

</pre>136.731 step3
<pre>runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step3_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

</pre>136.793 step3
<pre>runTheMatrix-results/136.793_RunDoubleEG2017C+RunDoubleEG2017C+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017/step3_RunDoubleEG2017C+RunDoubleEG2017C+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017.log

</pre>136.874 step3
<pre>runTheMatrix-results/136.874_RunEGamma2018C+RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM/step3_RunEGamma2018C+RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM.log

</pre>136.898 step3
<pre>runTheMatrix-results/136.898_RunParkingBPH2018B+RunParkingBPH2018B+HLTDR2_2018+RECODR2_2018reHLT_skimParkingBPH_Offline+HARVEST2018/step3_RunParkingBPH2018B+RunParkingBPH2018B+HLTDR2_2018+RECODR2_2018reHLT_skimParkingBPH_Offline+HARVEST2018.log

</pre>1306.0 step3
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step3_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>312.0 step3
<pre>runTheMatrix-results/312.0_Pyquen_ZeemumuJets_pt10_2760GeV_2021+Pyquen_ZeemumuJets_pt10_2760GeV_2021+DIGIHI2021MIX+RECOHI2021MIX+HARVESTHI2021PPRECO/step3_Pyquen_ZeemumuJets_pt10_2760GeV_2021+Pyquen_ZeemumuJets_pt10_2760GeV_2021+DIGIHI2021MIX+RECOHI2021MIX+HARVESTHI2021PPRECO.log

</pre>1304.182 step3
<pre>runTheMatrix-results/1304.182_ProdZEE_13UP18+ProdZEE_13UP18+DIGIUP18PROD1+RECOPRODUP18bParking+MINIAODMCUP18bParking/step3_ProdZEE_13UP18+ProdZEE_13UP18+DIGIUP18PROD1+RECOPRODUP18bParking+MINIAODMCUP18bParking.log

</pre>1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15.log

</pre>10042.0 step3
<pre>runTheMatrix-results/10042.0_ZMM_13+2017+ZMM_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_ZMM_13+2017+ZMM_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log

</pre>9.0 step3
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step3_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>25.0 step3
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>10024.0 step3
<pre>runTheMatrix-results/10024.0_TTbar_13+2017+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_TTbar_13+2017+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log

</pre>10824.0 step3
<pre>runTheMatrix-results/10824.0_TTbar_13+2018+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_TTbar_13+2018+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log

</pre>11634.0 step3
<pre>runTheMatrix-results/11634.0_TTbar_14TeV+2021+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA/step3_TTbar_14TeV+2021+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA.log

</pre>25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25.log

</pre>12434.0 step3
<pre>runTheMatrix-results/12434.0_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA/step3_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA.log

</pre>10224.0 step3
<pre>runTheMatrix-results/10224.0_TTbar_13+2017PU+TTbar_13TeV_TuneCUETP8M1_GenSim+DigiPU+RecoFakeHLTPU+HARVESTFakeHLTPU+Nano/step3_TTbar_13+2017PU+TTbar_13TeV_TuneCUETP8M1_GenSim+DigiPU+RecoFakeHLTPU+HARVESTFakeHLTPU+Nano.log

</pre>23234.0 step3
<pre>runTheMatrix-results/23234.0_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/step3_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log

</pre>250202.181 step4
<pre>runTheMatrix-results/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25/step4_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25.log

</pre>28234.0 step3
<pre>runTheMatrix-results/28234.0_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/step3_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log

</pre>23434.999 step4
<pre>runTheMatrix-results/23434.999_TTbar_14TeV+2026D49PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU/step4_TTbar_14TeV+2026D49PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py TTbar_8TeV_TuneCUETP8M1_cfi  --conditions auto:run1_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot Realistic8TeVCollision : FAILED - time: date Fri Dec  4 22:14:13 2020-date Fri Dec  4 22:11:23 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:Fake2,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_Fake2 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --processName=HLTRECO --filein file:RelVal_Raw_Fake2_DATA.root --fileout file:RelVal_Raw_Fake2_DATA_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:15:02 2020-date Fri Dec  4 22:11:31 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:PRef,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run3_data_PRef --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_PRef_DATA.root --fileout file:RelVal_Raw_PRef_DATA_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:17:51 2020-date Fri Dec  4 22:11:36 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:HIon,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run3_data_HIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3_pp_on_PbPb --processName=HLTRECO --filein file:RelVal_Raw_HIon_DATA.root --fileout file:RelVal_Raw_HIon_DATA_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:17:23 2020-date Fri Dec  4 22:11:37 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run3_mc_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_GRun_MC.root --fileout file:RelVal_Raw_GRun_MC_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:24:04 2020-date Fri Dec  4 22:11:40 2020 s - exit: 16640
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc_l1stage1 --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_25ns : FAILED - time: date Fri Dec  4 22:14:37 2020-date Fri Dec  4 22:11:41 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:PRef,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run3_mc_PRef --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_PRef_MC.root --fileout file:RelVal_Raw_PRef_MC_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:18:33 2020-date Fri Dec  4 22:11:45 2020 s - exit: 16640
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 : FAILED - time: date Fri Dec  4 22:15:01 2020-date Fri Dec  4 22:11:47 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run3_data_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_PIon_DATA.root --fileout file:RelVal_Raw_PIon_DATA_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:16:45 2020-date Fri Dec  4 22:11:50 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:Fake,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run1_mc_Fake --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --processName=HLTRECO --filein file:RelVal_Raw_Fake_MC.root --fileout file:RelVal_Raw_Fake_MC_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:17:55 2020-date Fri Dec  4 22:11:55 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:HIon,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run3_mc_HIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3_pp_on_PbPb --processName=HLTRECO --filein file:RelVal_Raw_HIon_MC.root --fileout file:RelVal_Raw_HIon_MC_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:19:58 2020-date Fri Dec  4 22:11:59 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:Fake2,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_Fake2 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --processName=HLTRECO --filein file:RelVal_Raw_Fake2_MC.root --fileout file:RelVal_Raw_Fake2_MC_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:18:26 2020-date Fri Dec  4 22:12:04 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:Fake1,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_Fake1 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_25ns --processName=HLTRECO --filein file:RelVal_Raw_Fake1_MC.root --fileout file:RelVal_Raw_Fake1_MC_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:18:21 2020-date Fri Dec  4 22:12:07 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:Fake,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run1_data_Fake --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --processName=HLTRECO --filein file:RelVal_Raw_Fake_DATA.root --fileout file:RelVal_Raw_Fake_DATA_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:16:08 2020-date Fri Dec  4 22:12:12 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run3_data_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_GRun_DATA.root --fileout file:RelVal_Raw_GRun_DATA_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:24:46 2020-date Fri Dec  4 22:12:15 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:Fake1,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_Fake1 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_25ns --processName=HLTRECO --filein file:RelVal_Raw_Fake1_DATA.root --fileout file:RelVal_Raw_Fake1_DATA_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:16:11 2020-date Fri Dec  4 22:12:18 2020 s - exit: 16640
cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run3_mc_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_PIon_MC.root --fileout file:RelVal_Raw_PIon_MC_HLT_RECO.root : FAILED - time: date Fri Dec  4 22:18:51 2020-date Fri Dec  4 22:12:21 2020 s - exit: 16640

slava77 commented:
> When I ran the RelVals I found an error in the following workflows:

I should've read the PR description more carefully, this requires a GT to be able to test (`We are waiting for the final GTs.`)

slava77 commented:
> > When I ran the RelVals I found an error in the following workflows:
> 
> I should've read the PR description more carefully, this requires a GT to be able to test (`We are waiting for the final GTs.`)

actually, @jainshilpi mentioned to me that this PR will have a temporary GT useful for tests

jainshilpi commented:
@slava77  All the requests to be included in GTs have been placed. In addition, is it possible to use the self created GT for these tests?

slava77 commented:
> @slava77 All the requests to be included in GTs have been placed. In addition, is it possible to use the self created GT for these tests?

yes, the GT can be updated in the autoCond.py file and can be included in this PR

bainbrid commented:
@jainshilpi @slava77 does that require action on my part?

jainshilpi commented:
@bainbrid since you have all the setup for the PR, I think its easier if you can change this file here: https://github.com/cms-sw/cmssw/blob/master/Configuration/AlCa/python/autoCond.py  (if its fine for you else let me know) @slava77 which tags should be updated? Is it all the ones mentioned here? https://github.com/cms-sw/cmssw/pull/32391#issuecomment-739030093
(run2_data_Fake2 etc?)

slava77 commented:
>  @slava77 which tags should be updated? Is it all the ones mentioned here?

run1, run2 data/MC (including run2 HI variety) as well as run3, phase2 MC GTs.
It's worth to push the autoCond.py update here only after `runTheMatrix.py -l limited -i all --ibeos` passes locally

bainbrid commented:
The updated autoConf.py is available in this PR: https://github.com/cms-sw/cmssw/pull/32421. I'll attempt to test locally overnight.

Also, we wish to update the ID model, as discussed here: https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/16#issuecomment-740812091

bainbrid commented:
There is a [comment](https://github.com/cms-sw/cmssw/pull/32372#discussion_r537897250) in the discussion for #32372 that [this](https://github.com/cms-sw/cmssw/pull/32372/commits/ee6ceb36b6b166d07d876ac9b27c12749a6ba3eb) change should be forward ported to this PR. 

I note that the `RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronID_cff.py ` file has been deprecated in the 11_X cycle and the appropriate default settings are defined once (and not overridden) in the addDescriptions method [here](https://github.com/cms-sw/cmssw/blob/aef1aa25e768afab60bf1f1102e25d79e0725c7a/RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc#L170-L180).

Is this satisfactory?

slava77 commented:
> Is this satisfactory?

I'm not sure I read the preceding text correctly; it reads like "no change will be made", is that so?

if we ever need to support modifier-based updates to lowPtGsfElectronID, the lowPtGsfElectronID_cff will have to be recovered; if not now, it will likely happen in the future.

bainbrid commented:
> I'm not sure I read the preceding text correctly; it reads like "no change will be made", is that so?

Yes, no change needed (currently).

bainbrid commented:
Running `runTheMatrix.py -l limited -i all --ibeos` locally with the PR #32421

bainbrid commented:
I ran the command `runTheMatrix.py -l limited -i all --ibeos` locally with the changes from PR #32096. 

My test involved modifying the autoConf.py file using these commands:
```
cd $CMSSW_BASE/src
git cms-addpkg Configuration/AlCa
git remote add christopheralanwest git@github.com:christopheralanwest/cmssw.git
git fetch christopheralanwest
git diff HEAD 89df990 -- Configuration/AlCa/python/autoCond.py >& patch.diff
git apply patch.diff
```

The tests finished with `34 33 29 20 14 4 1 1 1 tests passed, 1 0 4 1 0 0 0 0 0 failed`. How do I interpret this information?

I see the following non-zero exit codes and Exceptions, below.

```
5.1_TTbar+TTbarFS+HARVESTFS Step0-FAILED Step1-NOTRUN  - time date Tue Dec  8 21:28:06 2020-date Tue Dec  8 21:26:16 2020; exit: 16640 0

----- Begin Fatal Exception 08-Dec-2020 21:28:03 CET-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'FEVTDEBUGHLToutput_step'
   [2] Prefetching for module PoolOutputModule/'FEVTDEBUGHLToutput'
   [3] Prefetching for module LowPtGsfElectronIDProducer/'lowPtGsfElectronID'
   [4] Calling method for module LowPtGsfElectronFinalizer/'lowPtGsfElectrons'
Exception Message:
No data of type "GBRForestD" with label "lowPtElectron_eb_ecalOnly_05To50_mean" in record "GBRDWrapperRcd"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------

9.0_Higgs200ChargedTaus+Higgs200ChargedTausINPUT+DIGI+RECO+HARVEST Step0-PASSED Step1-PASSED Step2-FAILED Step3-NOTRUN  - time date Tue Dec  8 21:36:40 2020-da
te Tue Dec  8 21:29:35 2020; exit: 0 0 16640 0

----- Begin Fatal Exception 08-Dec-2020 21:36:36 CET-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  Event run: 1 lumi: 71 event: 7002 stream: 0
   [1] Running path 'dqmoffline_step'
   [2] Prefetching for module DQMMessageLogger/'DQMMessageLogger'
   [3] Prefetching for module LogErrorHarvester/'logErrorHarvester'
   [4] Prefetching for module LowPtGsfElectronIDProducer/'lowPtGsfElectronID'
   [5] Calling method for module LowPtGsfElectronFinalizer/'lowPtGsfElectrons'
Exception Message:
No data of type "GBRForestD" with label "lowPtElectron_eb_ecalOnly_05To50_mean" in record "GBRDWrapperRcd"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------

25.0_TTbar+TTbarINPUT+DIGI+RECOAlCaCalo+HARVEST+ALCATT Step0-PASSED Step1-PASSED Step2-FAILED Step3-NOTRUN Step4-NOTRUN  - time date Tue Dec  8 21:39:25 2020-d
ate Tue Dec  8 21:32:02 2020; exit: 0 0 16640 0 0

----- Begin Fatal Exception 08-Dec-2020 21:39:20 CET-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 4 stream: 0
   [1] Running path 'dqmoffline_step'
   [2] Prefetching for module DQMMessageLogger/'DQMMessageLogger'
   [3] Prefetching for module LogErrorHarvester/'logErrorHarvester'
   [4] Prefetching for module LowPtGsfElectronIDProducer/'lowPtGsfElectronID'
   [5] Calling method for module LowPtGsfElectronFinalizer/'lowPtGsfElectrons'
Exception Message:
No data of type "GBRForestD" with label "lowPtElectron_eb_ecalOnly_05To50_mean" in record "GBRDWrapperRcd"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------

23234.0_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+HARVESTGlobal Step0-PASSED Step1-PASSED Step2-FAILED Step3-NOTRUN  - time date Tue Dec  8 22:41:38 2020-date Tue Dec  8 22:34:00 2020; exit: 0 0 16640 0

----- Begin Fatal Exception 08-Dec-2020 22:41:32 CET-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  Event run: 1 lumi: 2 event: 108 stream: 0
   [1] Running path 'dqmoffline_1_step'
   [2] Prefetching for module DQMMessageLogger/'DQMMessageLogger'
   [3] Prefetching for module LogErrorHarvester/'logErrorHarvester'
   [4] Prefetching for module LowPtGsfElectronIDProducer/'lowPtGsfElectronID'
   [5] Calling method for module LowPtGsfElectronFinalizer/'lowPtGsfElectrons'
Exception Message:
No data of type "GBRForestD" with label "lowPtElectron_eb_ecalOnly_05To50_mean" in record "GBRDWrapperRcd"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------

23434.999_TTbar_14TeV+2026D49PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU Step0-PASSED Step1-PASSED Step2-PASSED Step3-FAILED Step4-NOTRUN  - time date Tue Dec  8 22:55:41 2020-date Tue Dec  8 22:40:17 2020; exit: 0 0 0 16640 0

----- Begin Fatal Exception 08-Dec-2020 22:55:30 CET-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  Event run: 1 lumi: 2 event: 108 stream: 0
   [1] Running path 'dqmoffline_1_step'
   [2] Prefetching for module DQMMessageLogger/'DQMMessageLogger'
   [3] Prefetching for module LogErrorHarvester/'logErrorHarvester'
   [4] Prefetching for module LowPtGsfElectronIDProducer/'lowPtGsfElectronID'
   [5] Calling method for module LowPtGsfElectronFinalizer/'lowPtGsfElectrons'
Exception Message:
No data of type "GBRForestD" with label "lowPtElectron_eb_ecalOnly_05To50_mean" in record "GBRDWrapperRcd"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------

28234.0_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal Step0-PASSED Step1-PASSED Step2-FAILED Step3-NOTRUN  - time date Tue Dec  8 22:59:28 2020-date Tue Dec  8 22:41:09 2020; exit: 0 0 16640 0

----- Begin Fatal Exception 08-Dec-2020 22:59:22 CET-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'dqmoffline_1_step'
   [2] Prefetching for module DQMMessageLogger/'DQMMessageLogger'
   [3] Prefetching for module LogErrorHarvester/'logErrorHarvester'
   [4] Prefetching for module LowPtGsfElectronIDProducer/'lowPtGsfElectronID'
   [5] Calling method for module LowPtGsfElectronFinalizer/'lowPtGsfElectrons'
Exception Message:
No data of type "GBRForestD" with label "lowPtElectron_eb_ecalOnly_05To50_mean" in record "GBRDWrapperRcd"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
```

@christopheralanwest @jainshilpi @cavalari

christopheralanwest commented:
@bainbrid The updated GTs in PR #32421 do not update Phase 2 or Run 1 MC and so the corresponding RelVals fail. Do you intend to disable the low pT regression for these scenarios or is there an appropriate payload that can be used for these GTs? In the data GTs, IOV range [1, 294645) is used for 2016 (and also includes Run 1), and IOV range [314472, Infinity) is used for 2018 (and later).

jainshilpi commented:
@christopheralanwest I just did what we had done for other electron regression in e/gamma at that time- i.e. using run number starting from 1. I was going to place request using 2016 tags for run1 and 2018 tags for phase II.  Otherwise we could just re-request the data tags with the proper run range for this.  What would you suggest?

jainshilpi commented:
@christopheralanwest for now I have created requests to queues: 112X_mcRun1_realistic_v1 (2016 tags) and 112X_mcRun4_realistic_v4 (phase II) - probably quickest solution for now?

christopheralanwest commented:
> @christopheralanwest I just did what we had done for other electron regression in e/gamma at that time- i.e. using run number starting from 1. I was going to place request using 2016 tags for run1 and 2018 tags for phase II. Otherwise we could just re-request the data tags with the proper run range for this. What would you suggest?

The data tags are fine. We don't know the run numbers for Phase 2 so it wouldn't be possible to create a data tag including Phase 2 IOVs even in principle.

I will defer to @slava77 for the decision (for both 10_6_X and master) regarding whether the low pT energy regression should be disabled for Run 1 or Phase 2 using era mechanism or whether we should add the tags to the GTs.

slava77 commented:
> I will defer to @slava77 for the decision (for both 10_6_X and master) regarding whether the low pT energy regression should be disabled for Run 1 or Phase 2 using era mechanism or whether we should add the tags to the GTs.

generally, we ask that the same algorithms/modules run in all generic workflows (with rather limited exceptions where costs are prohibitive or if some subdetector parts changed).

christopheralanwest commented:
I updated the Run 1 and Phase 2 GTs in PR #32421 and triggered the tests in that PR. It seems that when multiple PRs are tested together, the bot now only provides notifications from the PR in which the test was triggered.

christopheralanwest commented:
The tests of PR #32421 with this PR and cms-data/RecoEgamma-ElectronIdentification#16 completed successfully with changes in the comparison tests only in low pT electron distributions. Can I sign off on the autoCond update or is there a reason to wait until this PR is ready as well?

slava77 commented:
> Can I sign off on the autoCond update or is there a reason to wait until this PR is ready as well?

does the autoCond PR #32421 work by itself? 
If so, it sounds practical to sign it.

christopheralanwest commented:
> > Can I sign off on the autoCond update or is there a reason to wait until this PR is ready as well?
> 
> does the autoCond PR #32421 work by itself?
> If so, it sounds practical to sign it.

OK, great. I've signed the autoCond PR.

bainbrid commented:
I can confirm that all wfs, tested locally with `runTheMatrix.py -l limited -i all --ibeos`, run to completion for this PR when tested with the latest version of #32421. 

I will also update the ID model in due course, as indicated here: https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/16#issuecomment-740812091

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-384d8a/11511/summary.html
**CMSSW**: CMSSW_11_3_X_2020-12-10-1100/slc7_amd64_gcc900

cmsbuild commented:
Comparison results are now available
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-384d8a/11511/summary.html
**CMSSW**: CMSSW_11_3_X_2020-12-10-1100/slc7_amd64_gcc900

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-384d8a/1304.182_ProdZEE_13UP18+ProdZEE_13UP18+DIGIUP18PROD1+RECOPRODUP18bParking+MINIAODMCUP18bParking
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-384d8a/136.898_RunParkingBPH2018B+RunParkingBPH2018B+HLTDR2_2018+RECODR2_2018reHLT_skimParkingBPH_Offline+HARVEST2018

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2313 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2747014
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2746979
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 35 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 153 log files, 37 edm output root files, 36 DQM output files

slava77 commented on RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronFinalizer.cc:
```diff
@@ -0,0 +1,67 @@
+  edm::ParameterSetDescription psd;
+  psd.setUnknown();
+  desc.add<edm::ParameterSetDescription>("regressionConfig", psd);
+  descriptions.addDefault(desc);
```
```suggestion
  descriptions.addWithDefaultLabel(desc);
```
why not actually inject a reference config?

bainbrid replied:
Updated as indicated, although I don't understand what descriptions.addWithDefaultLabel does. 

> why not actually inject a reference config?

Do you mean add the default python config as hardcoded c++? I'm not sure how to add nested dicts, plus the regressionConfig is cloned from another config.

slava77 replied:
`addDefault` stays in c++ side, while `addWithDefaultLabel` creates a valid python config fragment, in this case in `$CMSSW_BASE/cfipython/$SCRAM_ARCH/RecoEgamma/EgammaElectronProducers/lowPtGsfElectronFinalizer_cfi.py`

slava77 commented on PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py:
```diff
@@ -69,11 +67,30 @@
+
+# (3) apply energy regression
+from RecoEgamma.EgammaElectronProducers.lowPtGsfElectronFinalizer_cfi import lowPtGsfElectrons
+lowPtGsfElectronsPostRegression = lowPtGsfElectrons.clone(previousGsfElectronsTag = "lowPtGsfElectrons")
+_makePatLowPtElectronsTask.add(lowPtGsfElectronsPostRegression)
+
+# Switch to energy-regressed electrons as input
+sourceElectronsPostRegression = "lowPtGsfElectronsPostRegression"
+run2_miniAOD_UL.toModify(lowPtElectronMatch, src = sourceElectronsPostRegression)
+run2_miniAOD_UL.toModify(patLowPtElectrons, electronSource = sourceElectronsPostRegression)
+run2_miniAOD_UL.toModify(rekeyLowPtGsfElectronSeedValueMaps, gsfElectrons = sourceElectronsPostRegression)
+run2_miniAOD_UL.toModify(lowPtGsfElectronID, electrons = sourceElectronsPostRegression)
```
```suggestion
_makePatLowPtElectronsTask.add(lowPtGsfElectrons)
```

if this is needed only in the context of run2_miniAOD_UL, why not simply use the same `lowPtGsfElectrons` module name?
you will need a `run2_miniAOD_UL.toModify(lowPtGsfElectrons, previousGsfElectronsTag = "lowPtGsfElectrons::@skipCurrentProcess")` in what is now `lowPtGsfElectronFinalizer_cfi`

bainbrid replied:
Done.

slava77 commented on PhysicsTools/PatAlgos/python/selectionLayer1/lowPtElectronSelector_cfi.py:
```diff
@@ -6,7 +6,7 @@
 selectedPatLowPtElectrons = cms.EDFilter("PATElectronSelector",
     src = cms.InputTag("patLowPtElectrons"),
-    cut = cms.string("pt>1. && electronID('ID')>1.5"),
+    cut = cms.string("pt>1. && electronID('ID')>0.0"),
```
is this the new 10% point matching the new training?

bainbrid replied:
Yes, although this has been updated again due to the use of a new (smaller) model in cms-data/RecoEgamma-ElectronIdentification#16. The new threshold is -0.25, which again provides a 10% mistag rate and ~90% signal efficiency.

slava77 commented on RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronFinalizer.cc:
```diff
@@ -0,0 +1,67 @@
+  for (auto const& electron : event.get(previousGsfElectrons_)) {
+    outputElectrons.emplace_back(electron);
+    auto& newElectron = outputElectrons.back();
+    if (regression_) {
+      regression_->modifyObject(newElectron);
+    }
```
```suggestion
    regression_->modifyObject(newElectron);
```
the preceding code implies the `regression_` is non-null

bainbrid replied:
Done.

slava77 commented on RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronFinalizer_cfi.py:
```diff
@@ -0,0 +1,45 @@
+    ),
+)
+
+lowPtGsfElectrons = cms.EDProducer("LowPtGsfElectronFinalizer",
```
normally the _cfi file name should match the module defined in it. I suggest to git mv this to `lowPtGsfElectrons_cfi.py`

Also, once the `LowPtGsfElectronIDProducer::fillDescriptions` is updated to inject a valid default config, please clone from it here.

bainbrid replied:
> Also, once the LowPtGsfElectronIDProducer::fillDescriptions is updated to inject a valid default config, please clone from it here.

I'm confused by this request. Do you mean IDProducer or Finalizer? If the latter, please see my comments here https://github.com/cms-sw/cmssw/pull/32391#discussion_r541454001

slava77 replied:
right, I meant the Finalizer

slava77 commented on RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectrons_cfi.py:
```diff
@@ -1,10 +1,10 @@
 from RecoEgamma.EgammaElectronProducers.gsfElectrons_cfi import ecalDrivenGsfElectrons
 
-lowPtGsfElectrons = ecalDrivenGsfElectrons.clone(gsfElectronCoresTag = "lowPtGsfElectronCores")
+lowPtGsfElectronsPreRegression = ecalDrivenGsfElectrons.clone(gsfElectronCoresTag = "lowPtGsfElectronCores")
```
the file should be renamed to `lowPtGsfElectronsPreRegression_cfi.py`

bainbrid replied:
Done.

slava77 commented on RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronFinalizer_cfi.py:
```diff
@@ -0,0 +1,45 @@
+
+from RecoEgamma.EgammaTools.regressionModifier_cfi import regressionModifier106XUL
+
+lowPtRegressionModifier = regressionModifier106XUL.clone(
```
```suggestion
_lowPtRegressionModifier = regressionModifier106XUL.clone(
```
if there is no need to export this generically outside of this file

bainbrid replied:
Done.

slava77 commented:
> DQMHistoTests: Total failures: 12

BTW, is there a plan to extend the standard DQM/Validation to cover the (slimmed)lowPtElectrons?

slava77 commented:
> Studies on the energy regression and ID can be found here and here.

both links are pointing to this PR itself.
Please check and update to correct links.

slava77 commented:
Looking at the jenkins outputs, the differences are in expected areas

In UL re-miniAOD on jetHT wf 136.88811 `_slimmedLowPtElectrons_` count is down by 20%
![all_OldVSNew_RunJetHT2018DreMINIAODULwf136p88811c_patElectrons_slimmedLowPtElectrons__PAT_obj___electronIDs__0__second](https://user-images.githubusercontent.com/4676718/101970435-32be2500-3bdf-11eb-80e9-6b8dc1bbe974.png)
these are background-like electrons; so, perhaps such a significant reduction after the updated preselection is fine

especially considering that in EGamma R2018C wf 136.874 essentially the same number of electrons passes
![all_mini_OldVSNew_RunEGamma2018Cwf136p874c_patElectrons_slimmedLowPtElectrons__reRECO_obj___electronIDs__0__second](https://user-images.githubusercontent.com/4676718/101970946-723a4080-3be2-11eb-834a-2565e76d46a3.png)


the post-regression energy error uncertainties are much larger
in  jetHT wf 136.88811
![all_OldVSNew_RunJetHT2018DreMINIAODULwf136p88811c_patElectrons_slimmedLowPtElectrons__PAT_obj_correctedEcalEnergyError](https://user-images.githubusercontent.com/4676718/101970504-a7915f00-3bdf-11eb-9e7f-5cc7bb822b80.png)

similarly also in the more signal-like EGamma R2018C wf 136.874 
![all_mini_OldVSNew_RunEGamma2018Cwf136p874c_patElectrons_slimmedLowPtElectrons__reRECO_obj_correctedEcalEnergyError](https://user-images.githubusercontent.com/4676718/101971003-e1179980-3be2-11eb-9573-22f1ec86aebc.png)

about 20% of the electrons have `r= correctedEcalEnergyError()/correctedEcalEnergy() > 0.2`, compared to just 1 in the baseline (40% vs 3% with `r>0.1`) .
Even for electrons with pt>20GeV the baseline has a median of `r` at  around 0.0125 vs this PR  at around 0.032.
Is this expected (is the error filled correctly)?
(here and in the baseline `correctedEcalEnergy` is the same by value as `ecalEnergy` ).

it would be nice to see this (relative) error value on properly signal events, perhaps even a pull distribution.

cavalari commented:
Hi Slava, 
I have checked the pulls and they seem correct to me. 
I have added slides 12 to 15 to my presentation in the EGM meeting during the CMS week
https://indico.cern.ch/event/879938/
with a few selected plots for the pull as requested,
for central barrel (low pt range and high pt range) and for outer endcaps (low pt range and high pt range). 
You can see one of the plots here in attachment for the central barrel, high pt range (40-50 GeV), for the category where the Lowptelectron is matching also a PF electron. Pink = Supercluster raw energy, Blue = ECAL only regression, Black = GSF track p mode, Red regression of the combination of ECAL regressed energy and TK P mode. 
I think that the ecal energy error for LowPtElectrons, before the regression is applied (pink), is not as accurate as that after the regression (blue). So it is ok that they differ before and after regression. 
Cheers,
Francesca

![pull_bin_011103](https://user-images.githubusercontent.com/7326190/102097069-bcf7ca80-3e25-11eb-9da2-78945cb87180.png)

slava77 commented:
> Hi Slava,
> I have checked the pulls and they seem correct to me.

Thank you for the check and the plots.
It looks like we are down to just the somewhat technical code review follow up here now.

bainbrid commented:
@slava77 yes I'll get to the review this week - thanks for the feedback. 

Just a heads up that we would also like to update the ID model to one that gives very comparable performance but a significant reduction in size. We'll have to retune the score threshold correspondingly.

slava77 commented:
> @slava77 yes I'll get to the review this week - thanks for the feedback.

Thank you.

One more thing to follow up is to show the performance costs; please provide the timing and memory use for this update.

slava77 commented:
> @slava77 yes I'll get to the review this week - thanks for the feedback.

just checking if the follow up will happen this week or in January.

bainbrid committed:
slava's review

bainbrid committed:
retuned threshold for new 2020Nov15 model with depth10

bainbrid commented:
> > DQMHistoTests: Total failures: 12
> 
> BTW, is there a plan to extend the standard DQM/Validation to cover the (slimmed)lowPtElectrons?

Yes, @jainshilpi and I are in touch and plan to provide this in the new year.

bainbrid commented:
> > Studies on the energy regression and ID can be found here and here.
> 
> both links are pointing to this PR itself.
> Please check and update to correct links.

Fixed.

bainbrid commented:
> > Hi Slava,
> > I have checked the pulls and they seem correct to me.
> 
> Thank you for the check and the plots.
> It looks like we are down to just the somewhat technical code review follow up here now.

@slava77 I've acted on your review comments, there are a couple of questions.

bainbrid commented:
> @slava77 yes I'll get to the review this week - thanks for the feedback.
> 
> Just a heads up that we would also like to update the ID model to one that gives very comparable performance but a significant reduction in size. We'll have to retune the score threshold correspondingly.

This has been done, with cms-data/RecoEgamma-ElectronIdentification#16 updated. The discussion associated with the cms-data PR highlights the switch to a model that uses the depth=10 hyperparameter, which reduces the performance only by a marginal amount, yet reduces the footprint significantly from 39MB --> 8MB. 

All recent code and model changes have been tested locally with wfs 136.88811 and 1325.518.

bainbrid commented:
> > @slava77 yes I'll get to the review this week - thanks for the feedback.
> 
> Thank you.
> 
> One more thing to follow up is to show the performance costs; please provide the timing and memory use for this update.

I'll try to do this today.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32391/20550

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronFinalizer_cfi.py:
    - Added: 9b1d484cd92ecf88c6e0307422103274855b1aa1
    - Modified: c771cfdbe865ecf6825495ebf57f67d3e37f0064, 711f80d26cc20afe4341457d69b826930b1a555b
    - Deleted: 3fd5cce0edea899fde86224ba247595399ed0b52

cmsbuild commented:
Pull request #32391 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

bainbrid commented:
**Timing, memory, and footprint studies:**

Performed with the 136.88811 (JETHT 2018D) workflow, re-running step2 with the command `cmsDriver.py step2 --conditions auto:run2_data_relval -s PAT,DQM:@miniAODDQM --process PAT -n 100 --data --era Run2_2018 --eventcontent MINIAOD,DQM --procModifiers run2_miniAOD_UL_preSummer20 --scenario pp --datatier MINIAOD,DQMIO --filein filelist:step1_dasquery.log --lumiToProcess step1_lumiRanges.log --fileout file:step2.root --customise Validation/Performance/TimeMemoryInfo.py --no_exec --python_filename step2_TEST.py`

I show results for:
1) the default PAT sequence in CMSSW_11_3_X_2020-12-17-1100, labelled "_MASTER" below, and 
2) the modified PAT sequence according to this PR, labelled "_PR" below. 

_Memory and CPU:_

```
~/tools/getTimeMemSummary.sh step2_MASTER.log
Summary for 1000 events
Max VSIZ 3412.95 on evt 900 ; max RSS 2142.59 on evt 977
Time av 0.812608 s/evt   max 15.5743 s on evt 1
CPU av 0 s/evt   max 0 s on evt -1
M1 Time av 0.797831 s/evt   max 3.01303 s on evt 167
M1 CPU av 0 s/evt   max 0 s on evt -1
M8 Time av 0.797254 s/evt   max 3.01303 s on evt 167
M16 Time av 0.797723 s/evt   max 3.01303 s on evt 167

~/tools/getTimeMemSummary.sh step2_PR.log
Summary for 1000 events
Max VSIZ 3472.48 on evt 900 ; max RSS 1982.52 on evt 963
Time av 0.819054 s/evt   max 17.0869 s on evt 1
CPU av 0 s/evt   max 0 s on evt -1
M1 Time av 0.80277 s/evt   max 3.04201 s on evt 328
M1 CPU av 0 s/evt   max 0 s on evt -1
M8 Time av 0.802069 s/evt   max 3.04201 s on evt 328
M16 Time av 0.802601 s/evt   max 3.04201 s on evt 328
```

_Timing report:_

```
~/tools/timeDiffFromReport.sh step2_MASTER.log step2_PR.log
Couldn't parse time report using CPU and wall-clock format: trying Wall-clock only
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.23%         0.00 ms/ev ->         1.62 ms/ev lowPtGsfElectrons
   -1.626943      -0.02%         0.17 ms/ev ->         0.02 ms/ev rekeyLowPtGsfElectronSeedValueMaps
   +1.076923      +0.00%         0.01 ms/ev ->         0.03 ms/ev triggerMatchMonitor_miniAOD
   -0.856348      -0.61%         7.18 ms/ev ->         2.88 ms/ev lowPtGsfElectronID
   +0.381775      +0.03%         0.52 ms/ev ->         0.76 ms/ev electronMVAValueMapProducer
  ---------- ------------     --------                  ----       ------------
Job total:  0.701113 s/ev ==> 0.702135 s/ev

The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.23%         0.00 ms/ev ->         1.61 ms/ev lowPtGsfElectrons
   -1.609127      -0.02%         0.17 ms/ev ->         0.02 ms/ev rekeyLowPtGsfElectronSeedValueMaps
   +1.016072      +0.00%         0.01 ms/ev ->         0.03 ms/ev triggerMatchMonitor_miniAOD
   -0.854404      -0.62%         7.17 ms/ev ->         2.88 ms/ev lowPtGsfElectronID
   +0.070755      +0.50%        47.32 ms/ev ->        50.79 ms/ev MINIAODoutput
  ---------- ------------     --------                  ----       ------------
Job total:  0.696364 s/ev ==> 0.698252 s/ev
```

_Comparing products and sizes:_

```
~/tools/compareProducts.sh step2_MASTER.root step2_PR.root _ 1 1
Checking process _ step2_MASTER.root and step2_PR.root with useUnpacked=no (if above 1 or 1%):
Compare packed values
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch
-----------------------------------------------------------------
   1293.8 ->      1560.5        267     18.7   0.43     patElectrons_slimmedLowPtElectrons__PAT.
-------------------------------------------------------------
    61640 ->       61906        267             0.4     ALL BRANCHES
```

bainbrid commented:
> > @slava77 yes I'll get to the review this week - thanks for the feedback.
> 
> just checking if the follow up will happen this week or in January.

@slava77 sorry for the delay in replying, starting tackling this yesterday, but took longer than expected.

bainbrid commented:
@slava77 I've attached a slide that provides details on how correlated the two BDT model scores are, and the effect on the 2020Nov28 performance after the application of a threshold of 1.5 on the score from the 2020Sept15 model.

These studies would be pertinent to e.g. a "UL nanoAOD v9" production based on the UL re-miniAOD samples.

[201223_ID_correlation.pdf](https://github.com/cms-sw/cmssw/files/5734502/201223_ID_correlation.pdf)

slava77 commented:
> [201223_ID_correlation.pdf](https://github.com/cms-sw/cmssw/files/5734502/201223_ID_correlation.pdf)

please add this to the PR description.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-384d8a/11868/summary.html
**CMSSW**: CMSSW_11_3_X_2020-12-24-1100/slc7_amd64_gcc900


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-384d8a/1304.182_ProdZEE_13UP18+ProdZEE_13UP18+DIGIUP18PROD1+RECOPRODUP18bParking+MINIAODMCUP18bParking
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-384d8a/1325.81_TTbar_13_106Xv1NanoAODINPUT+TTbar_13_106Xv1NanoAODINPUT+NANOEDMMC2017_106XMiniAODv1+HARVESTNANOAODMC2017_106XMiniAODv1
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-384d8a/136.8523_RunJetHT2018C_nanoULremini+RunJetHT2018C_nanoULremini+NANOEDM2018_106Xv2+HARVESTNANOAOD2018_106Xv2
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-384d8a/136.898_RunParkingBPH2018B+RunParkingBPH2018B+HLTDR2_2018+RECODR2_2018reHLT_skimParkingBPH_Offline+HARVEST2018

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2416 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2716967
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2716935
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

slava77 commented on RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronsPreRegression_cfi.py:
```diff
@@ -0,0 +1,10 @@
+
+from Configuration.Eras.Modifier_fastSim_cff import fastSim
+fastSim.toModify(lowPtGsfElectronsPreRegression,ctfTracksTag = "generalTracksBeforeMixing")
+
+from Configuration.ProcessModifiers.pp_on_AA_cff import pp_on_AA
+pp_on_AA.toModify(lowPtGsfElectronsPreRegression.preselection, minSCEtBarrel = 15.0)
+pp_on_AA.toModify(lowPtGsfElectronsPreRegression.preselection, minSCEtEndcaps = 15.0)
```
```suggestion
```
not necessary, the origin `ecalDrivenGsfElectrons.clone` already has these applied

bainbrid replied:
Done.

slava77 commented on RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectrons_cfi.py:
```diff
@@ -1,10 +1,48 @@
-from Configuration.ProcessModifiers.pp_on_AA_cff import pp_on_AA
-pp_on_AA.toModify(lowPtGsfElectrons.preselection, minSCEtBarrel = 15.0)
-pp_on_AA.toModify(lowPtGsfElectrons.preselection, minSCEtEndcaps = 15.0)
+lowPtGsfElectrons = cms.EDProducer("LowPtGsfElectronFinalizer",
+                                   previousGsfElectronsTag = cms.InputTag("lowPtGsfElectronsPreRegression"),
+                                   regressionConfig = _lowPtRegressionModifier,
+)
```
following https://github.com/cms-sw/cmssw/pull/32391#discussion_r541460175

```suggestion
from RecoEgamma.EgammaElectronProducers.lowPtGsfElectronFinalizer_cfi import lowPtGsfElectronFinalizer
lowPtGsfElectrons = lowPtGsfElectronFinalizer.clone(
                                   previousGsfElectronsTag = "lowPtGsfElectronsPreRegression",
                                   regressionConfig = _lowPtRegressionModifier,
)
```

bainbrid replied:
Done.

slava77 commented:
compared to the previous iteration in https://github.com/cms-sw/cmssw/pull/32391#issuecomment-743692605

in jetHT (136.88811) now the number of passing electrons is about the same, 3/110 more (it was down by 20% before)
![all_OldVSNew_RunJetHT2018DreMINIAODULwf136p88811c_patElectrons_slimmedLowPtElectrons__PAT_obj___electronIDs__0__second](https://user-images.githubusercontent.com/4676718/103141042-964a5700-46a3-11eb-87ef-42f4e1c2b3a5.png)

and in events with more real electrons in EGamma Run2018 (wf 136.874) is similarly very slightly up by 5/111
![all_mini_OldVSNew_RunEGamma2018Cwf136p874c_patElectrons_slimmedLowPtElectrons__reRECO_obj___electronIDs__0__second](https://user-images.githubusercontent.com/4676718/103141072-0f49ae80-46a4-11eb-8559-91dbdc863261.png)
This looks reasonable.


I looked at memory use in a bit more detail (more than just MemoryCheck printouts)
https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_11_3_X_2020-12-23-2300-orig.136.88811.step2.IgProf.1.MEM_LIVE
vs https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_11_3_X_2020-12-23-2300-sign1115.136.88811.step2.IgProf.1.MEM_LIVE
shows about 19MiB decrease in memory use, where the leading contributions are from 46MiB reduction in LowPtGsfElectronIDProducer (new training model), and a 27MiB increase from extra GBRDWrapperRcd coming via LowPtGsfElectronFinalizer.
This looks reasonable.

slava77 commented:
@bainbrid 
please clarify on the status of following up on the latest review comments.
Thank you.

bainbrid commented:
Hi @slava77 Thanks for the feedback. I will tackle them in full tomorrow.

bainbrid commented:
> > [201223_ID_correlation.pdf](https://github.com/cms-sw/cmssw/files/5734502/201223_ID_correlation.pdf)
> 
> please add this to the PR description.

I've updated the PR description.

bainbrid committed:
slava: responding to review comments

bainbrid commented:
> @bainbrid
> please clarify on the status of following up on the latest review comments.
> Thank you.

@slava77 sorry for the delay. Rebased to the latest IB. This commit https://github.com/cms-sw/cmssw/pull/32391/commits/edee328474a7c46c5e3f4f73a5ba874803bd202c is the only relevant commit, in response to your latest round of review comments. Reran 25202.0 without problems.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32391/20673

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronFinalizer_cfi.py:
    - Added: 42d5b2252d6be670b5a22061c4a4edf96497b3ac
    - Modified: 21e0d3181e53c4259101c35fc1debd18365a657b, ccac8ec4c030493b0e4b594c4cb32d0f73fdf912
    - Deleted: 9365d7d46472234477ab900eef4407439d1a46d2

cmsbuild commented:
Pull request #32391 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-384d8a/12031/summary.html
**COMMIT**: edee328474a7c46c5e3f4f73a5ba874803bd202c
**CMSSW**: CMSSW_11_3_X_2021-01-08-1100/slc7_amd64_gcc900


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-384d8a/1304.182_ProdZEE_13UP18+ProdZEE_13UP18+DIGIUP18PROD1+RECOPRODUP18bParking+MINIAODMCUP18bParking

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2412 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2716967
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2716944
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

slava77 commented:
@cmsbuild please test

in case the error was transient; 
the addOn fastsim2 failed in `TBLayer::groupedCompatibleDetsV <- .. <- fastsim::TrackerSimHitProducer::interact`
which is unrelated to this PR

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-384d8a/12038/summary.html
**COMMIT**: edee328474a7c46c5e3f4f73a5ba874803bd202c
**CMSSW**: CMSSW_11_3_X_2021-01-08-1100/slc7_amd64_gcc900


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-384d8a/1304.182_ProdZEE_13UP18+ProdZEE_13UP18+DIGIUP18PROD1+RECOPRODUP18bParking+MINIAODMCUP18bParking

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2415 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2716967
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2716938
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/32391	edee328
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show differences only in the low-pt electron energy, ID, and related distributions as expected; some plots and technical analysis was posted earlier in https://github.com/cms-sw/cmssw/pull/32391#issuecomment-751284274

this PR requires cms-data/RecoEgamma-ElectronIdentification#16

slava77 commented:
@silviodonato @qliphy 
please consider bypassing the analysis signature
Thank you.

silviodonato commented:
merge

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
